### ISSUE TITLE
Virtual (headless) outputs 

### ISSUE THREAD

**Issue by TRPB** [WayfireWM/wayfire#1710]
wlroots supports `wlr_headless_add_output` which allows creating a headless output. This is useful for a number of use-cases:

1. Headless VNC access
2. Using another device as a second monitor  ( There's a sample sway set up in this reddit post: https://www.reddit.com/r/swaywm/comments/k1zl41/thank_you_devs_free_ipad_repurposed_as_a_second/ )
3. Or why I want this: **Configuring a real monitor to mirror a virtual output so that windows aren't moved around when a screen is turned off** (I know there is a plugin for this, but it's intermittent at best and if you VNC into a machine after turning off a display all the windows are now squashed onto one monitor)

Sway currently supports this so there is an example implementation in the PRs listed here: https://github.com/swaywm/sway/issues/5553#issuecomment-711135069 

## Suggested implementation options

Add a `headless = true` option to an output:

```
[output:HEADLESS-1]
mode = 3840x2160@60000
headless = true

[output:DP-1]
mode = mirror HEADLESS-1
```

Alternatively, and this limits it to my use case (vnc and retaining screen positions in a tidier way) a `permanent` flag on a specified output:

```
[output:DP-1]
mode = 3840x2160@60000
permanent = true
```

Which would specify that the output exists even if the monitor is disconnected. The back end could be implemented exactly as above: a virtual output that always exists, mirrored by the physical display when it's connected

**Comment by soreau** [comment]
> wlroots supports `wlr_headless_add_output` which allows creating a headless output. This is useful for a number of use-cases:
> 
>     1. Headless VNC access
> 
>     2. Using another device as a second monitor  ( There's a sample sway set up in this reddit post: https://www.reddit.com/r/swaywm/comments/k1zl41/thank_you_devs_free_ipad_repurposed_as_a_second/ )
> 
>     3. Or why I want this: **Configuring a real monitor to mirror a virtual output so that windows aren't moved around when a screen is turned off** (I know there is a plugin for this, but it's intermittent at best and if you VNC into a machine after turning off a display all the windows are now squashed onto one monitor)


Aren't 1. and 2. basically the same thing? You should be able to do this since wlroots supports it through [environment variables](https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/docs/env_vars.md):

Put this in wayfire.ini `[input-device:wlr_virtual_pointer_v1] output = HEADLESS-1` and start wayfire with the env vars:
`WLR_RENDER_DRM_DEVICE=/dev/dri/renderD128 WLR_BACKENDS=drm,headless wayfire`
Start wayvnc:
`wayvnc -o HEADLESS-1 localhost 5900`
Start a vnc viewer client:
`vinagre localhost:5900`

This should start wayfire with a headless output in addition to drm, that you can use vnc with.

About 3., I think the existing plugin should be improved or fixed, instead of trying to work around it as you describe.

**Comment by TRPB** [comment]
Thanks for the suggestion.

I tried that, unfortunately the mouse and  keyboard events no longer work, including switching TTY. Wayfire hasn't frozen as my clock keeps ticking but as I can't switch TTY or exit wayfire, I can't see the logs.

**Comment by soreau** [comment]
Can you run wayfire with `-d` for debug, in addition to the other options, redirect the output to file with `wayfire -d &> $HOME/wayfire.log` or the like, and post the uploaded log here?

**Comment by TRPB** [comment]
Thanks:

[wayfire.log](https://github.com/WayfireWM/wayfire/files/10582571/wayfire.log)

It does show `[types/output/cursor.c:410] Falling back to software cursor on output 'HEADLESS-1'`

**Comment by soreau** [comment]
I see what you mean. Here, `WLR_BACKENDS=headless,libinput,drm` seems to work. Apparently you need to add libinput and ordering is important too.

**Comment by TRPB** [comment]
Thanks. This actually works for my use case and works really well. You do also need to specify the number of headless outputs with `WLR_HEADLESS_OUTPUTS=1` for this to work and I had to use `drm,headless,libinput` for some reason. 

I got random crashes before (around 25% of the time) when disconnecting/reconnecting a monitor. By mirroring a virtual monitor:

1. The windows never get rearranged because the headless output is still connected (big win. I was looking at getting a KVM without EDID monitor emulation and this is a nice workaround)
2. I haven't had a crash when disconnecting/connecting the display!

I have no idea how this works with things like Adaptive Sync if the frame is being rendered on a virtual display then copied I guess it wouldn't sync properly? Due to a quirk of my monitor at first I had the monitor running at 30hz and the headless display at 60hz. Running glxgears, it's synced to 60hz so when doing this, definitely need to make sure the headless display is configured exactly the same as the actual display.

For anyone who's interested, here's my setup. Launch variables:

```
WLR_RENDER_DRM_DEVICE=/dev/dri/renderD128 WLR_HEADLESS_OUTPUTS=1  WLR_BACKENDS=drm,headless,libinput wayfire
```

`wayfire.ini`:

```
[output:DP-2]
mode = mirror HEADLESS-1

[output:HEADLESS-1]
mode = 3840x2160@60000
```

I'm happy for this ticket to be closed, this niche use case is possible without changes to wayfire.
